x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java

```diff
@@ -52,4 +71,134 @@ public void putConnector(Connector connector, ActionListener<DocWriteResponse> l
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+
+        try {
+            clientWithOrigin.delete(
+                deleteRequest,
+                new DelegatingIndexNotFoundActionListener<>(connectorId, listener, (l, deleteResponse) -> {
+                    if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                        l.onFailure(new ResourceNotFoundException(connectorId));
+                        return;
+                    }
+                    l.onResponse(deleteResponse);
+                })
+            );
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+
+    }
+
+    /**
+     * List the {@link Connector} in ascending order of their ids.
+     *
+     * @param from From index to start the search from.
+     * @param size The maximum number of {@link Connector}s to return.
+     * @param listener The action listener to invoke on response/failure.
+     */
+    public void listConnectors(int from, int size, ActionListener<ConnectorIndexService.ConnectorResult> listener) {
+        try {
+            final SearchSourceBuilder source = new SearchSourceBuilder().from(from)
+                .size(size)
+                .query(new MatchAllQueryBuilder())
+                .fetchSource(true)
+                .sort(Connector.ID_FIELD.getPreferredName(), SortOrder.ASC);
+            final SearchRequest req = new SearchRequest(CONNECTOR_INDEX_NAME).source(source);
+            clientWithOrigin.search(req, new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse searchResponse) {
+                    listener.onResponse(mapSearchResponseToConnectorList(searchResponse));
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof IndexNotFoundException) {
+                        listener.onResponse(new ConnectorIndexService.ConnectorResult(Collections.emptyList(), 0L));
+                        return;
+                    }
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private static ConnectorIndexService.ConnectorResult mapSearchResponseToConnectorList(SearchResponse response) {
+        final List<Connector> connectorResults = Arrays.stream(response.getHits().getHits())
+            .map(ConnectorIndexService::hitToConnector)
+            .toList();
+        return new ConnectorIndexService.ConnectorResult(connectorResults, (int) response.getHits().getTotalHits().value);
+    }
+
+    private static Connector hitToConnector(SearchHit searchHit) {
+
+        // todo: don't return sensitive data from configuration in list endpoint
+
+        return Connector.fromXContentBytes(searchHit.getSourceRef(), XContentType.JSON);
```

Comment:
Related to our previous conversation, is this something that needs to be done, or can we just return the source directly without parsing it?

Reply:
There was a request to obfuscate some user-provided secrets stored in the index (e.g. api keys to authenticate against a datasource). 

Given that - I needed to implement parses for all of "helper" connector data structures (for parsing update endpoints) - I decided that it might be elegant to:
1. Parse the whole connector document
2. Apply some internal function to obfuscate user secrets before returning data to the user

Note: the obfuscation logic is not implemented yet

Reply:
Got it, thanks!